### PR TITLE
Avoid using union in `crypto/bio/bss_dgram.c`

### LINKED ISSUE #18637: Strict aliasing concerns in bss_dgram.c
The `crypto/bio/bss_dgram.c` file has a number of strict aliasing problems.  There are a number of definitions of a union:
```c
    union {
        size_t s;
        int i;
    }
```
and a definition:
```c
    union {
        unsigned __int64 ul;
        FILETIME ft;
    }
```
The concern is that the address is taken of a union member and this is passed to a function.  This breaks the strict aliasing rules.  These instances ought to be rewritten using casts.

### PR DESCRIPTION
**Pull request by xtkoba**
Fixes #18637

### CODE REVIEW

**Review by thesamesam — COMMENTED** [review]
Looks much better.

File: crypto/bio/bss_dgram.c [hunk 1]
```diff
@@ -1918,20 +1900,19 @@ static void get_current_time(struct timeval *t)
...
-        unsigned __int64 ul;
-        FILETIME ft;
-    } now;
+    unsigned __int64 now_ul;
+    FILETIME now_ft;
 
     GetSystemTime(&st);
-    SystemTimeToFileTime(&st, &now.ft);
+    SystemTimeToFileTime(&st, &now_ft);
+    memcpy(&now_ul, &now_ft, sizeof(unsigned __int64));
```
Comment (paulidale) [inline crypto/bio/bss_dgram.c:1908]: Wouldn't:
```c
now_ul = ((unsigned __int64)not_ft.dwHighDateTime << 32) + not_ft.dwLowDateTime;
```
be safer?  I know it's ever so slightly different to what the existing code did but it's endian friendly.

Comment (xtkoba) [reply crypto/bio/bss_dgram.c]: Fixed. Thanks!

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Does this count as a bug?
It might be relevant for 3.0 depending on the compiler.

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Yep, bug.

**Comment by paulidale** [comment]
Merged to both, thanks for the contribution.

**Pull request closed without merging**